pkg/storage/utils/eosfs/eosfs.go

```diff
@@ -537,22 +543,257 @@ func (fs *eosfs) UnsetArbitraryMetadata(ctx context.Context, ref *provider.Refer
+	}
+	return resInfo.PermissionSet.InitiateFileUpload, nil
+}
+
+func (fs *eosfs) hasReadAccess(ctx context.Context, user *userpb.User, ref *provider.Reference) (bool, error) {
+	resInfo, err := fs.GetMD(ctx, ref, nil)
+	if err != nil {
+		return false, err
+	}
+	return resInfo.PermissionSet.InitiateFileDownload, nil
+}
+
+func encodeLock(l *provider.Lock) (string, error) {
+	data, err := json.Marshal(l)
+	if err != nil {
+		return "", err
+	}
+	return b64.StdEncoding.EncodeToString(data), nil
+}
+
+func decodeLock(raw string) (*provider.Lock, error) {
+	data, err := b64.StdEncoding.DecodeString(raw)
+	if err != nil {
+		return nil, err
+	}
+	l := new(provider.Lock)
+	err = json.Unmarshal(data, l)
+	if err != nil {
+		return nil, err
+	}
+	return l, nil
 }
 
 // RefreshLock refreshes an existing lock on the given reference
-func (fs *eosfs) RefreshLock(ctx context.Context, ref *provider.Reference, lock *provider.Lock) error {
-	return errtypes.NotSupported("unimplemented")
+func (fs *eosfs) RefreshLock(ctx context.Context, ref *provider.Reference, newLock *provider.Lock) error {
+	// TODO (gdelmont): check if the new lock is already expired?
+
+	if newLock.Type == provider.LockType_LOCK_TYPE_SHARED {
+		return errtypes.NotSupported("shared lock not yet implemented")
+	}
+
+	oldLock, err := fs.GetLock(ctx, ref)
+	if err != nil {
+		switch err.(type) {
+		case errtypes.NotFound:
+			// the lock does not exist
+			return errtypes.BadRequest("file was not locked")
+		default:
+			return err
+		}
+	}
+
+	user, err := getUser(ctx)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: error getting user")
+	}
+
+	// check if the user hold the old lock
+	if oldLock.User != nil && !utils.UserEqual(oldLock.User, user.Id) {
+		return errtypes.BadRequest("caller does not hold the lock")
+	}
+
+	// check if the holder is the same of the new lock
+	if !sameHolder(oldLock, newLock) {
+		return errtypes.BadRequest("caller does not hold the lock")
+	}
+
+	// the cs3apis require to have the write permission on the resource
+	// to set a lock
+	has, err := fs.hasWriteAccess(ctx, user, ref)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: cannot check if user has write access on resource")
+	}
+	if !has {
+		return errtypes.BadRequest("user has not write access on resource")
+	}
+
+	// set the xattr with the new value
+	path, err := fs.resolve(ctx, ref)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: error resolving reference")
+	}
+	path = fs.wrap(ctx, path)
+
+	auth, err := fs.getUserAuth(ctx, user, path)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: error getting uid and gid for user")
+	}
+
+	encodedLock, err := encodeLock(newLock)
+	if err != nil {
+		return errors.Wrap(err, "eosfs: error encoding lock")
+	}
+
+	attr := &eosclient.Attribute{
+		Type: UserAttr,
+		Key:  LockKeyAttr,
+		Val:  encodedLock,
+	}
+
+	return fs.c.SetAttr(ctx, auth, attr, false, false, path)
+}
+
+func sameHolder(l1, l2 *provider.Lock) bool {
+	same := true
```

Comment:
```suggestion
	same := false
```

Reply:
Why false? Because the user and appname in the lock are optional, and must be checked if not empty, if they are either empty or nil, they have the same "holder" (i.e. no one)

Reply:
ah right, forget that then! Let me double check that we enforce somewhere that either the user or the appname are set on `SetLock`...